### PR TITLE
feat: Add LRU hash

### PR DESCRIPTION
**Pull request by diptanshumittal**

### CODE REVIEW

File: gapic-common/lib/gapic/lru_hash.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        new_node = Node.new key, value
+        insert_at_top new_node
+        @cache[key] = new_node
+      end
+    end
+
+    private
+
+    def move_to_top node
+      return if node == @start
```
Comment (dazuma) [inline gapic-common/lib/gapic/lru_hash.rb:52]: `node.equal? @start` better expresses this intent. `==` tests value equality and could be true for two distinct objects with the same semantic value. `equal?` tests object identity which is what we're looking for here. Same a couple lines down.

File: gapic-common/lib/gapic/lru_hash.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        node.next = @start
+        @start.prev = node
+        @start = node
+      end
+    end
+
+    ##
+    # Node class for linked list.
+    class Node
+      attr_accessor :key
```
Comment (dazuma) [inline gapic-common/lib/gapic/lru_hash.rb:88]: I don't think the `Node#key` field is ever used. Can we get rid of it?

**Review by dazuma — CHANGES_REQUESTED** [review]

File: gapic-common/lib/gapic/lru_hash.rb (hunk 1 above)
Comment (diptanshumittal) [reply gapic-common/lib/gapic/lru_hash.rb]: Thanks for the explanation. Updated the code to use `.equals?`

File: gapic-common/lib/gapic/lru_hash.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Gapic
+  ##
+  # Linked list based hash maintaining the order of
```
[the remark this answers is not in the document]
Comment (diptanshumittal) [reply gapic-common/lib/gapic/lru_hash.rb:19]: Marked the class `private`.

File: gapic-common/lib/gapic/lru_hash.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        @end = node
+      else
+        node.next = @start
+        @start.prev = node
+        @start = node
+      end
+    end
+
+    ##
+    # Node class for linked list.
```
[the remark this answers is not in the document]
Comment (diptanshumittal) [reply gapic-common/lib/gapic/lru_hash.rb:90]: Marked the class `private`.

File: gapic-common/lib/gapic/lru_hash.rb (hunk 2 above)
Comment (diptanshumittal) [reply gapic-common/lib/gapic/lru_hash.rb:93]: Currently, the `key` attr in `Node` inner class is used for deleting the <key, value> pair from the hash. [Link](https://github.com/googleapis/gapic-generator-ruby/blob/368ee2425a091328b89ab61d6aa3ffb04aeef4fa/gapic-common/lib/gapic/lru_hash.rb#L71)
Without this, we won't have a direct reference for which <key, value> pair to delete from the hash.

Comment (dazuma) [reply gapic-common/lib/gapic/lru_hash.rb:93]: Ah yes, you're right, I missed that.

**Review by dazuma — APPROVED** [review]

**Pull request merged**